File: src/main/java/me/hhitt/disasters/game/drop/ItemDropManager.java
package me.hhitt.disasters.game.drop;

import me.hhitt.disasters.Disasters;
import me.hhitt.disasters.arena.Arena;
import me.hhitt.disasters.model.drop.ItemDrop;
import me.hhitt.disasters.storage.file.Configuration;
import me.hhitt.disasters.storage.file.FileManager;
import me.hhitt.disasters.util.Msg;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.World;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.ArmorStand;
import org.bukkit.entity.Item;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.PotionMeta;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.potion.PotionType;
import org.bukkit.util.Vector;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Random;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;

public final class ItemDropManager {

    public static final class ItemEntry {
        public final Material material;
        public final int amount;
        public final PotionType potionType;

        public ItemEntry(Material material, int amount, PotionType potionType) {
            this.material = material;
            this.amount = amount;
            this.potionType = potionType;
        }

        public ItemEntry(Material material, int amount) {
            this(material, amount, null);
        }

        public ItemEntry(Material material) {
            this(material, 1, null);
        }
    }

    public static final class ItemTier {
        public final String name;
        public final int chance;
        public final String color;
        public final List<ItemEntry> items;

        public ItemTier(String name, int chance, String color, List<ItemEntry> items) {
            this.name = name;
            this.chance = chance;
            this.color = color;
            this.items = items;
        }
    }

    public static final NamespacedKey DROP_KEY = new NamespacedKey(Disasters.getInstance(), "disaster_drop");
    public static final NamespacedKey SPECIAL_DROP_KEY = new NamespacedKey(Disasters.getInstance(), "special_drop");

    private static final ConcurrentHashMap<Arena, CopyOnWriteArrayList<ItemDrop>> activeDrops = new ConcurrentHashMap<Arena, CopyOnWriteArrayList<ItemDrop>>();

    private static boolean enabled = true;
    private static int dropRate = 20;
    private static int maxDropsPerCycle = 10;
    private static int despawnTime = 30;
    private static int spawnRadius = 15;
    private static boolean glow = true;
    private static final List<ItemTier> tiers = new ArrayList<ItemTier>();

    private static final Random RANDOM = new Random();

    private ItemDropManager() {
    }

    public static void loadConfig() {
        Configuration mainConfig = FileManager.get("config");
        boolean globalEnabled = mainConfig != null && mainConfig.getBoolean("enable-item-drops", true);

        Configuration config = FileManager.get("item-drops");
        if (config == null) {
            enabled = false;
            return;
        }

        enabled = globalEnabled && config.getBoolean("enabled", true);
        dropRate = config.getInt("drop-rate", 20);
        maxDropsPerCycle = config.getInt("max-players-per-cycle", 10);
        despawnTime = config.getInt("despawn-time", 30);
        spawnRadius = config.getInt("spawn-radius", 15);
        glow = config.getBoolean("glow", true);

        tiers.clear();
        ConfigurationSection tiersSection = config.getConfigurationSection("tiers");
        if (tiersSection != null) {
            for (String tierName : tiersSection.getKeys(false)) {
                ConfigurationSection tierSection = tiersSection.getConfigurationSection(tierName);
                if (tierSection == null) {
                    continue;
                }
                int chance = tierSection.getInt("chance", 0);
                String color = tierSection.getString("color", "<white>");
                List<String> rawItems = tierSection.getStringList("items");
                List<ItemEntry> items = new ArrayList<ItemEntry>();
                for (String raw : rawItems) {
                    ItemEntry entry = parseItemEntry(raw);
                    if (entry != null) {
                        items.add(entry);
                    }
                }
                if (!items.isEmpty()) {
                    tiers.add(new ItemTier(tierName, chance, color, items));
                }
            }
        }

        if (enabled && tiers.isEmpty()) {
            Disasters.getInstance().getLogger().warning("Item drops enabled but no valid tiers configured!");
        }
    }

    private static ItemEntry parseItemEntry(String raw) {
        String[] parts = raw.split(":");
        String materialName = parts[0].toUpperCase(Locale.ROOT);
        Material material = Material.matchMaterial(materialName);

        if (material == null) {
            Disasters.getInstance().getLogger().warning("Invalid item drop material: " + materialName);
            return null;
        }

        if (material == Material.POTION || material == Material.SPLASH_POTION || material == Material.LINGERING_POTION) {
            PotionType potionType = null;
            if (parts.length > 1) {
                try {
                    potionType = parsePotionType(parts[1].toUpperCase(Locale.ROOT));
                } catch (Exception e) {
                    Disasters.getInstance().getLogger().warning("Invalid potion type: " + parts[1]);
                }
            }
            return new ItemEntry(material, 1, potionType);
        }

        int amount = 1;
        if (parts.length > 1) {
            try {
                amount = Integer.parseInt(parts[1]);
            } catch (NumberFormatException e) {
                amount = 1;
            }
        }
        return new ItemEntry(material, amount);
    }

    private static PotionType parsePotionType(String name) {
        if ("INSTANT_HEAL".equals(name) || "HEALING".equals(name)) {
            return PotionType.HEALING;
        } else if ("FIRE_RESISTANCE".equals(name)) {
            return PotionType.FIRE_RESISTANCE;
        } else if ("REGENERATION".equals(name)) {
            return PotionType.REGENERATION;
        } else if ("SWIFTNESS".equals(name) || "SPEED".equals(name)) {
            return PotionType.SWIFTNESS;
        } else if ("SLOW_FALLING".equals(name)) {
            return PotionType.SLOW_FALLING;
        } else if ("RESISTANCE".equals(name)) {
            try {
                return PotionType.valueOf("STRONG_TURTLE_MASTER");
            } catch (IllegalArgumentException e) {
                return null;
            }
        } else {
            try {
                return PotionType.valueOf(name);
            } catch (IllegalArgumentException e) {
                return null;
            }
        }
    }

    private static ItemTier pickRandomTier() {
        if (tiers.isEmpty()) {
            return null;
        }
        int totalWeight = 0;
        for (ItemTier tier : tiers) {
            totalWeight += tier.chance;
        }
        if (totalWeight <= 0) {
            return tiers.get(RANDOM.nextInt(tiers.size()));
        }
        int roll = RANDOM.nextInt(totalWeight);
        for (ItemTier tier : tiers) {
            roll -= tier.chance;
            if (roll < 0) {
                return tier;
            }
        }
        return tiers.get(tiers.size() - 1);
    }

    public static void pulse(Arena arena, int gameTime) {
        if (!enabled || tiers.isEmpty()) {
            return;
        }

        CopyOnWriteArrayList<ItemDrop> drops = activeDrops.computeIfAbsent(arena, k -> new CopyOnWriteArrayList<ItemDrop>());

        java.util.Iterator<ItemDrop> iterator = drops.iterator();
        while (iterator.hasNext()) {
            ItemDrop drop = iterator.next();
            if (gameTime - drop.getSpawnTime() >= despawnTime || drop.getItem().isDead()) {
                drop.remove();
                drops.remove(drop);
            }
        }

        if (gameTime > 0 && gameTime % dropRate == 0) {
            List<Player> players = new ArrayList<Player>(arena.getAlive());
            Collections.shuffle(players, RANDOM);
            int count = Math.min(maxDropsPerCycle, players.size());
            for (int i = 0; i < count; i++) {
                spawnDrop(arena, gameTime, players.get(i));
            }
        }
    }

    private static void spawnDrop(Arena arena, int gameTime, Player player) {
        Location location = getRandomSurfaceLocation(arena, player);
        if (location == null) {
            return;
        }
        ItemTier tier = pickRandomTier();
        if (tier == null) {
            return;
        }
        ItemEntry entry = tier.items.get(RANDOM.nextInt(tier.items.size()));
        ItemStack itemStack = createItemStack(entry);
        String displayName = getDisplayName(entry);

        World world = location.getWorld();
        if (world == null) {
            return;
        }

        Item item = world.dropItem(location, itemStack);
        item.setVelocity(new Vector(0, 0, 0));
        item.setUnlimitedLifetime(true);
        item.setCustomNameVisible(false);
        item.getPersistentDataContainer().set(DROP_KEY, PersistentDataType.BYTE, (byte) 1);
        item.setCanPlayerPickup(true);
        item.setCanMobPickup(false);
        if (glow) {
            item.setGlowing(true);
        }

        Location labelLoc = location.clone().add(0.0, 1.2, 0.0);
        ArmorStand label = world.spawn(labelLoc, ArmorStand.class, stand -> {
            stand.setVisible(false);
            stand.setCustomNameVisible(true);
            stand.customName(Msg.parse(tier.color + displayName));
            stand.setGravity(false);
            stand.setMarker(true);
            stand.setSmall(true);
            stand.setInvulnerable(true);
            stand.setCanTick(false);
            stand.getPersistentDataContainer().set(DROP_KEY, PersistentDataType.BYTE, (byte) 1);
        });

        ItemDrop drop = new ItemDrop(item, label, gameTime);
        activeDrops.computeIfAbsent(arena, k -> new CopyOnWriteArrayList<ItemDrop>()).add(drop);
    }

    private static ItemStack createItemStack(ItemEntry entry) {
        ItemStack stack = new ItemStack(entry.material, entry.amount);
        if (entry.potionType != null) {
            if (stack.getItemMeta() instanceof PotionMeta) {
                PotionMeta meta = (PotionMeta) stack.getItemMeta();
                meta.setBasePotionType(entry.potionType);
                stack.setItemMeta(meta);
            }
        }
        if (stack.getItemMeta() != null) {
            org.bukkit.inventory.meta.ItemMeta meta = stack.getItemMeta();
            if (entry.material == Material.TNT) {
                meta.setDisplayName(net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer.plainText().serialize(Msg.parse("<red>Throwable TNT")));
                meta.getPersistentDataContainer().set(SPECIAL_DROP_KEY, PersistentDataType.BYTE, (byte) 1);
            } else if (entry.material == Material.WATER_BUCKET || entry.material == Material.MILK_BUCKET) {
                meta.getPersistentDataContainer().set(SPECIAL_DROP_KEY, PersistentDataType.BYTE, (byte) 1);
            }
            stack.setItemMeta(meta);
        }
        return stack;
    }

    private static String getDisplayName(ItemEntry entry) {
        if (entry.potionType != null) {
            String potionName = entry.potionType.name().replace("_", " ").toLowerCase(Locale.ROOT);
            potionName = potionName.substring(0, 1).toUpperCase(Locale.ROOT) + potionName.substring(1);
            String typeName;
            if (entry.material == Material.SPLASH_POTION) {
                typeName = "Splash";
            } else if (entry.material == Material.LINGERING_POTION) {
                typeName = "Lingering";
            } else {
                typeName = "";
            }
            String prefix = !typeName.isEmpty() ? typeName + " " : "";
            return prefix + "Potion of " + potionName;
        }
        if (entry.amount > 1) {
            String name = entry.material.name().replace("_", " ").toLowerCase(Locale.ROOT);
            name = name.substring(0, 1).toUpperCase(Locale.ROOT) + name.substring(1);
            return name + " x" + entry.amount;
        }
        String name = entry.material.name().replace("_", " ").toLowerCase(Locale.ROOT);
        return name.substring(0, 1).toUpperCase(Locale.ROOT) + name.substring(1);
    }

    private static Location getRandomSurfaceLocation(Arena arena, Player player) {
        World world = player.getWorld();
        Location playerLoc = player.getLocation();

        for (int i = 0; i < 10; i++) {
            int offsetX = RANDOM.nextInt(spawnRadius * 2 + 1) - spawnRadius;
            int offsetZ = RANDOM.nextInt(spawnRadius * 2 + 1) - spawnRadius;
            int x = playerLoc.getBlockX() + offsetX;
            int z = playerLoc.getBlockZ() + offsetZ;

            Location checkLoc = new Location(world, (double) x, playerLoc.getY(), (double) z);
            if (!arena.getBorderService().isLocationInArena(checkLoc)) {
                continue;
            }

            Double surfaceY = findSurface(world, x, z, playerLoc.getBlockY() - 10, playerLoc.getBlockY() + 10);
            if (surfaceY != null) {
                return new Location(world, x + 0.5, surfaceY + 1.0, z + 0.5);
            }
        }
        return null;
    }

    private static Double findSurface(World world, int x, int z, int minY, int maxY) {
        for (int y = maxY; y >= minY; y--) {
            org.bukkit.block.Block block = world.getBlockAt(x, y, z);
            org.bukkit.block.Block above = world.getBlockAt(x, y + 1, z);
            if (block.getType().isSolid() && above.getType().isAir()) {
                return (double) y;
            }
        }
        return null;
    }

    public static void onPickup(Arena arena, Item item) {
        CopyOnWriteArrayList<ItemDrop> drops = activeDrops.get(arena);
        if (drops == null) {
            return;
        }
        for (ItemDrop drop : drops) {
            if (drop.getItem().equals(item)) {
                drop.getLabel().remove();
                drops.remove(drop);
                return;
            }
        }
    }

    public static void clearDrops(Arena arena) {
        CopyOnWriteArrayList<ItemDrop> drops = activeDrops.remove(arena);
        if (drops != null) {
            for (ItemDrop drop : drops) {
                drop.remove();
            }
        }
    }

    public static void clearAll() {
        for (CopyOnWriteArrayList<ItemDrop> drops : activeDrops.values()) {
            for (ItemDrop drop : drops) {
                drop.remove();
            }
        }
        activeDrops.clear();
    }

    public static void cleanOrphanedEntities(Arena arena) {
        World world = arena.getCorner1().getWorld();
        if (world == null) {
            return;
        }
        for (org.bukkit.entity.Entity entity : world.getEntities()) {
            if (entity.getPersistentDataContainer().has(DROP_KEY, PersistentDataType.BYTE)) {
                if (arena.getBorderService().isLocationInArena(entity.getLocation())) {
                    entity.remove();
                }
            }
        }
    }
}

File: src/main/kotlin/me/hhitt/disasters/game/drop/ItemDropManager.kt
package me.hhitt.disasters.game.drop

import me.hhitt.disasters.Disasters
import me.hhitt.disasters.arena.Arena
import me.hhitt.disasters.model.drop.ItemDrop
import me.hhitt.disasters.util.Msg
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.World
import org.bukkit.entity.ArmorStand
import org.bukkit.entity.Item
import org.bukkit.inventory.ItemStack
import org.bukkit.util.Vector
import org.bukkit.inventory.meta.PotionMeta
import org.bukkit.persistence.PersistentDataType
import org.bukkit.potion.PotionType
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import kotlin.random.Random

/**
 * Manages item drop spawning, tracking, and cleanup for all arenas.
 */
object ItemDropManager {

    val DROP_KEY by lazy { NamespacedKey(Disasters.getInstance(), "disaster_drop") }

    private val activeDrops = ConcurrentHashMap<Arena, CopyOnWriteArrayList<ItemDrop>>()

    private var enabled = true
    private var dropRate = 20
    private var maxDropsPerCycle = 10
    private var despawnTime = 30
    private var spawnRadius = 15
    private var glow = true
    private var tiers = mutableListOf<ItemTier>()

    /**
     * Represents a parsed item entry from config.
     */
    data class ItemEntry(
        val material: Material,
        val amount: Int = 1,
        val potionType: PotionType? = null
    )

    /**
     * Represents a rarity tier with a weighted chance, display color, and item pool.
     */
    data class ItemTier(
        val name: String,
        val chance: Int,
        val color: String,
        val items: List<ItemEntry>
    )

    fun loadConfig() {
        // Global toggle from config.yml
        val mainConfig = me.hhitt.disasters.storage.file.FileManager.get("config")
        val globalEnabled = mainConfig?.getBoolean("enable-item-drops", true) ?: true

        val config = Disasters.getInstance().let { plugin ->
            val file = java.io.File(plugin.dataFolder, "item-drops.yml")
            if (!file.exists()) {
                plugin.saveResource("item-drops.yml", false)
            }
            org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(file)
        }

        enabled = globalEnabled && config.getBoolean("enabled", true)
        dropRate = config.getInt("drop-rate", 20)
        maxDropsPerCycle = config.getInt("max-players-per-cycle", 10)
        despawnTime = config.getInt("despawn-time", 30)
        spawnRadius = config.getInt("spawn-radius", 15)
        glow = config.getBoolean("glow", true)

        tiers.clear()
        val tiersSection = config.getConfigurationSection("tiers")
        if (tiersSection != null) {
            for (tierName in tiersSection.getKeys(false)) {
                val tierSection = tiersSection.getConfigurationSection(tierName) ?: continue
                val chance = tierSection.getInt("chance", 0)
                val color = tierSection.getString("color", "<white>") ?: "<white>"
                val rawItems = tierSection.getStringList("items")
                val items = rawItems.mapNotNull { parseItemEntry(it) }
                if (items.isNotEmpty()) {
                    tiers.add(ItemTier(tierName, chance, color, items))
                }
            }
        }

        if (enabled && tiers.isEmpty()) {
            Disasters.getInstance().logger.warning("Item drops enabled but no valid tiers configured!")
        }
    }

    private fun parseItemEntry(raw: String): ItemEntry? {
        // Formats: MATERIAL, MATERIAL:AMOUNT, POTION:EFFECT, SPLASH_POTION:EFFECT
        val parts = raw.split(":")
        val materialName = parts[0].uppercase()
        val material = Material.matchMaterial(materialName)

        if (material == null) {
            Disasters.getInstance().logger.warning("Invalid item drop material: $materialName")
            return null
        }

        // Potion types
        if (material == Material.POTION || material == Material.SPLASH_POTION || material == Material.LINGERING_POTION) {
            val potionType = if (parts.size > 1) {
                try {
                    parsePotionType(parts[1].uppercase())
                } catch (e: Exception) {
                    Disasters.getInstance().logger.warning("Invalid potion type: ${parts[1]}")
                    null
                }
            } else null
            return ItemEntry(material, 1, potionType)
        }

        // Regular items with optional amount
        val amount = if (parts.size > 1) parts[1].toIntOrNull() ?: 1 else 1
        return ItemEntry(material, amount)
    }

    private fun parsePotionType(name: String): PotionType? {
        return when (name) {
            "INSTANT_HEAL", "HEALING" -> PotionType.HEALING
            "FIRE_RESISTANCE" -> PotionType.FIRE_RESISTANCE
            "REGENERATION" -> PotionType.REGENERATION
            "SWIFTNESS", "SPEED" -> PotionType.SWIFTNESS
            "SLOW_FALLING" -> PotionType.SLOW_FALLING
            "RESISTANCE" -> {
                try { PotionType.valueOf("STRONG_TURTLE_MASTER") } catch (_: Exception) { null }
            }
            else -> try { PotionType.valueOf(name) } catch (_: Exception) { null }
        }
    }

    /**
     * Picks a random tier based on configured chances (weighted random).
     */
    private fun pickRandomTier(): ItemTier? {
        if (tiers.isEmpty()) return null
        val totalWeight = tiers.sumOf { it.chance }
        if (totalWeight <= 0) return tiers.random()
        var roll = Random.nextInt(totalWeight)
        for (tier in tiers) {
            roll -= tier.chance
            if (roll < 0) return tier
        }
        return tiers.last()
    }

    /**
     * Called every second from GameTimer. Handles spawning and despawning.
     */
    fun pulse(arena: Arena, gameTime: Int) {
        if (!enabled || tiers.isEmpty()) return

        val drops = activeDrops.getOrPut(arena) { CopyOnWriteArrayList() }

        // Despawn expired drops
        val iterator = drops.iterator()
        while (iterator.hasNext()) {
            val drop = iterator.next()
            if (gameTime - drop.spawnTime >= despawnTime || drop.item.isDead) {
                drop.remove()
                drops.remove(drop)
            }
        }

        // Spawn a drop near each alive player (up to max-players-per-cycle)
        if (gameTime > 0 && gameTime % dropRate == 0) {
            val players = arena.alive.shuffled().take(maxDropsPerCycle)
            for (player in players) {
                spawnDrop(arena, gameTime, player)
            }
        }
    }

    private fun spawnDrop(arena: Arena, gameTime: Int, player: org.bukkit.entity.Player) {
        val location = getRandomSurfaceLocation(arena, player) ?: return
        val tier = pickRandomTier() ?: return
        val entry = tier.items.random()
        val itemStack = createItemStack(entry)
        val displayName = getDisplayName(entry)

        val world = location.world ?: return

        // Spawn the item entity with no velocity so it stays in place
        val item = world.dropItem(location, itemStack)
        item.velocity = Vector(0, 0, 0)
        item.setUnlimitedLifetime(true)
        item.isCustomNameVisible = false
        item.persistentDataContainer.set(DROP_KEY, PersistentDataType.BYTE, 1.toByte())
        item.setCanPlayerPickup(true)
        item.setCanMobPickup(false)
        if (glow) item.isGlowing = true

        // Spawn floating text label above the item with tier color
        val label = world.spawn(location.clone().add(0.0, 1.2, 0.0), ArmorStand::class.java) { stand ->
            stand.isVisible = false
            stand.isCustomNameVisible = true
            stand.customName(Msg.parse("${tier.color}$displayName"))
            stand.setGravity(false)
            stand.isMarker = true
            stand.isSmall = true
            stand.isInvulnerable = true
            stand.setCanTick(false)
            stand.persistentDataContainer.set(DROP_KEY, PersistentDataType.BYTE, 1.toByte())
        }

        val drop = ItemDrop(item, label, gameTime)
        activeDrops.getOrPut(arena) { CopyOnWriteArrayList() }.add(drop)
    }

    private fun createItemStack(entry: ItemEntry): ItemStack {
        val stack = ItemStack(entry.material, entry.amount)
        if (entry.potionType != null) {
            val meta = stack.itemMeta as? PotionMeta
            if (meta != null) {
                meta.basePotionType = entry.potionType
                stack.itemMeta = meta
            }
        }
        return stack
    }

    private fun getDisplayName(entry: ItemEntry): String {
        if (entry.potionType != null) {
            val potionName = entry.potionType.name.replace("_", " ").lowercase()
                .replaceFirstChar { it.uppercase() }
            val typeName = when (entry.material) {
                Material.SPLASH_POTION -> "Splash"
                Material.LINGERING_POTION -> "Lingering"
                else -> ""
            }
            val prefix = if (typeName.isNotEmpty()) "$typeName " else ""
            return "${prefix}Potion of $potionName"
        }
        if (entry.amount > 1) {
            val name = entry.material.name.replace("_", " ").lowercase()
                .replaceFirstChar { it.uppercase() }
            return "$name x${entry.amount}"
        }
        return entry.material.name.replace("_", " ").lowercase()
            .replaceFirstChar { it.uppercase() }
    }

    /**
     * Finds a surface spot near the given player within spawnRadius.
     */
    private fun getRandomSurfaceLocation(arena: Arena, player: org.bukkit.entity.Player): Location? {
        val world = player.world
        val playerLoc = player.location

        // Try a few times to find a valid surface spot near the player
        repeat(10) {
            val offsetX = Random.nextInt(-spawnRadius, spawnRadius + 1)
            val offsetZ = Random.nextInt(-spawnRadius, spawnRadius + 1)
            val x = playerLoc.blockX + offsetX
            val z = playerLoc.blockZ + offsetZ

            // Make sure the spot is within the arena
            val checkLoc = Location(world, x.toDouble(), playerLoc.y, z.toDouble())
            if (!arena.borderService.isLocationInArena(checkLoc)) return@repeat

            // Scan for surface near the player's Y level
            val surfaceY = findSurface(world, x, z, playerLoc.blockY - 10, playerLoc.blockY + 10)
            if (surfaceY != null) {
                return Location(world, x + 0.5, surfaceY + 1.0, z + 0.5)
            }
        }
        return null
    }

    /**
     * Scans downward from maxY to find the highest solid block with air above.
     */
    private fun findSurface(world: World, x: Int, z: Int, minY: Int, maxY: Int): Double? {
        for (y in maxY downTo minY) {
            val block = world.getBlockAt(x, y, z)
            val above = world.getBlockAt(x, y + 1, z)
            if (block.type.isSolid && above.type.isAir) {
                return y.toDouble()
            }
        }
        return null
    }

    /**
     * Called when a player picks up a disaster drop item.
     * Removes the associated label ArmorStand.
     */
    fun onPickup(arena: Arena, item: Item) {
        val drops = activeDrops[arena] ?: return
        val drop = drops.find { it.item == item } ?: return
        drop.label.remove()
        drops.remove(drop)
    }

    /**
     * Removes all active drops for an arena (called on game end).
     */
    fun clearDrops(arena: Arena) {
        val drops = activeDrops.remove(arena) ?: return
        drops.forEach { it.remove() }
    }

    /**
     * Removes all drops across all arenas (called on plugin disable).
     */
    fun clearAll() {
        activeDrops.forEach { (_, drops) -> drops.forEach { it.remove() } }
        activeDrops.clear()
    }

    /**
     * Removes any orphaned drop entities left in the arena (e.g. from a server crash).
     * Called when a new game starts in an arena.
     */
    fun cleanOrphanedEntities(arena: Arena) {
        val world = arena.corner1.world ?: return
        world.entities.forEach { entity ->
            if (entity.persistentDataContainer.has(DROP_KEY, PersistentDataType.BYTE)) {
                if (arena.borderService.isLocationInArena(entity.location)) {
                    entity.remove()
                }
            }
        }
    }
}
